src/measure_theory/lp_space.lean

```diff
@@ -1218,4 +1217,432 @@ by { rw eq, exact continuous_pos_part.comp continuous_neg }
+  cases (h_tendsto 1 ennreal.zero_lt_one ennreal.one_lt_top) with N h_tendsto_1,
+  specialize h_tendsto_1 N (le_refl N),
+  dsimp only at h_tendsto_1,
+  have h_add : f_lim = f_lim - f N + f N, by abel,
+  rw h_add,
+  refine lt_of_le_of_lt (snorm'_add_le (f_lim_meas.ae_measurable.sub (hf N).1) (hf N).1 _) _,
+  { rwa ennreal.to_real_of_real (zero_le_one.trans hp1), },
+  rw ennreal.add_lt_top,
+  split,
+  { refine lt_of_lt_of_le _ (@le_top ennreal _ 1),
+    have h_neg : f_lim - f N = -(f N - f_lim), by simp,
+    rwa [h_neg, snorm'_neg, ennreal.to_real_of_real (zero_le_one.trans hp1)], },
+  { have h_snorm_lt_top := (hf N).2,
+    rwa snorm_eq_snorm' hp_ne_zero ennreal.of_real_ne_top at h_snorm_lt_top, },
+end
+
+lemma cauchy_complete_ℒp_of_ne_top [complete_space E] [hp : fact(1 ≤ p)] (hp_ne_top : p ≠ ∞)
+  {f : ℕ → α → E} (hf : ∀ n, mem_ℒp (f n) p μ) {B : ℕ → ℝ≥0∞} (hB : ∑' i, B i < ∞)
+  (h_cau : ∀ (N n m : ℕ), N ≤ n → N ≤ m → snorm (f n - f m) p μ < B N) :
+  ∃ (f_lim : α → E) (hf_lim_meas : mem_ℒp f_lim p μ),
+    ∀ ε, 0 < ε → ε < ∞ → (∃ (N : ℕ), ∀ (n : ℕ), N ≤ n → snorm (f n - f_lim) p μ < ε) :=
+begin
+  rw ←ennreal.of_real_to_real hp_ne_top at hf,
+  have hp1 : 1 ≤ p.to_real,
+  { rw [← ennreal.of_real_le_iff_le_to_real hp_ne_top, ennreal.of_real_one],
+    exact hp.elim, },
+  have h_cau' : ∀ (N n m : ℕ), N ≤ n → N ≤ m → snorm' (f n - f m) (p.to_real) μ < B N,
+  { intros N n m hn hm,
+    specialize h_cau N n m hn hm,
+    rwa snorm_eq_snorm' (ennreal.zero_lt_one.trans_le hp.elim).ne.symm hp_ne_top at h_cau, },
+  obtain ⟨f_lim, h_lim_meas, h_cauchy⟩ := cauchy_complete_ℒp_of_ne_top' hf hp1 hB h_cau',
+  rw ennreal.of_real_to_real hp_ne_top at h_lim_meas,
+  refine ⟨f_lim, h_lim_meas, λ ε hε_pos hε_ne_top, _⟩,
+  obtain ⟨N, h⟩ := h_cauchy ε hε_pos hε_ne_top,
+  refine ⟨N, λ n hn, _⟩,
+  rw snorm_eq_snorm' (ennreal.zero_lt_one.trans_le hp.elim).ne.symm hp_ne_top,
+  exact h n hn,
+end
+
+lemma cauchy_complete_ℒp_top [complete_space E] {f : ℕ → α → E} (hf : ∀ n, mem_ℒp (f n) ∞ μ)
+  {B : ℕ → ℝ≥0∞} (hB : ∑' i, B i < ∞)
+  (h_cau : ∀ (N n m : ℕ), N ≤ n → N ≤ m → snorm (f n - f m) ∞ μ < B N) :
+  ∃ (f_lim : α → E) (hf_lim_ℒp : mem_ℒp f_lim ∞ μ),
+    ∀ (ε : ℝ≥0∞), 0 < ε → ε < ∞ → (∃ (N : ℕ), ∀ (n : ℕ), N ≤ n → snorm (f n - f_lim) ∞ μ < ε) :=
+begin
+  obtain ⟨f_lim, h_lim_meas, h_tendsto⟩ := cauchy_limit_ℒp_top hf hB h_cau,
+  refine ⟨f_lim, ⟨h_lim_meas.ae_measurable, _⟩, h_tendsto⟩,
+  cases (h_tendsto 1 ennreal.zero_lt_one ennreal.one_lt_top) with N h_tendsto_1,
+  specialize h_tendsto_1 N (le_refl N),
+  dsimp only at h_tendsto_1,
+  have h_add : f_lim = f_lim - f N + f N, by abel,
+  rw h_add,
+  refine lt_of_le_of_lt (snorm_add_le (h_lim_meas.ae_measurable.sub (hf N).1) (hf N).1 le_top) _,
+  rw ennreal.add_lt_top,
+  split,
+  { refine lt_of_lt_of_le _ (@le_top ennreal _ 1),
+    have h_neg : f_lim - f N = -(f N - f_lim), by simp,
+    rwa [h_neg, snorm_neg], },
+  { exact (hf N).2, },
+end
```

Comment:
I think this argument (deducing that the limit is in `L^p` when you already have the norm control) is the same for both finite and infinite `p`, and can probably be factored out (no need to give it twice!).

Reply:
It is the same, up to the real/ennreal juggling in the non-top case. Together `cauchy_complete_ℒp_of_ne_top'` and `cauchy_complete_ℒp_of_ne_top` correspond to `cauchy_complete_ℒp_top`, but with real/ennreal conversions in the mix. I'll rework them to isolate that argument from the conversions between real and ennreal.

Reply:
Yes, I was thinking it should be possible to do the juggling earlier (just after cauchy_limit_ℒp) to factor out the rest of the argument.

Reply:
It was in fact not hard to refactor. I've added a lemma `mem_ℒp_of_cauchy_limit`, and I use it in both finite and infinite cases.